TooltipDecorator/useTooltip.js

```diff
@@ -0,0 +1,266 @@
+	} = props;
+	const rtl = useI18nContext()?.rtl;
+
+	const [showing, setShowing] = useState(false);
+	const [layoutInfo, setLayoutInfo] = useState({
+		arrowAnchor: null,
+		labelOffset: null,
+		position: {top: 0, left: 0},
+		tooltipDirection: null
+	});
+	const mutableRef = useRef({
+		mutationObserver: null,
+		resizeObserver: null,
+		showTooltipJob: null,
+		setTooltipLayoutJob: null
+	});
+	const clientRef = useRef(null);
+	const tooltipRef = useRef(null);
+
+	const showTooltip = useCallback((client) => {
+		if (tooltipText) {
+			clientRef.current = client;
+			currentTooltip = client;
+			mutableRef.current.showTooltipJob?.startAfter(tooltipDelay);
+			mutableRef.current.mutationObserver?.observe(clientRef.current, {attributes: true, childList: true});
+			mutableRef.current.resizeObserver?.observe(clientRef.current);
+		}
+	}, [clientRef, tooltipDelay, tooltipText]);
+
+	const hideTooltip = useCallback(() => {
+		if (tooltipText) {
+			mutableRef.current.mutationObserver?.disconnect();
+			mutableRef.current.resizeObserver?.disconnect();
+
+			clientRef.current = null;
+			currentTooltip = null;
+
+			mutableRef.current.showTooltipJob?.stop();
+			mutableRef.current.setTooltipLayoutJob?.stop();
+
+			setShowing(false);
+		}
+	}, [clientRef, tooltipText]);
+
+	const startTooltipLayoutJob = useCallback(() => {
+		mutableRef.current.setTooltipLayoutJob?.startAfter(tooltipUpdateDelay);
+	}, [tooltipUpdateDelay]);
+
+	const handlers = useHandlers({
+		// Recalculate tooltip layout on keydown to make sure tooltip is positioned correctly in case something changes as a result of the keydown.
+		onKeyDown: handle(
+			forward('onKeyDown'),
+			not(forProp('disabled', true)),
+			() => {
+				startTooltipLayoutJob();
+			}
+		),
+		onMouseOver: handle(
+			forward('onMouseOver'),
+			forProp('disabled', true),
+			(ev) => {
+				showTooltip(ev.currentTarget);
+			}
+		),
+		onMouseOut: handle(
+			forward('onMouseOut'),
+			forProp('disabled', true),
+			(ev) => {
+				if (clientRef.current && !clientRef.current.contains(ev.relatedTarget)) {
+					hideTooltip();
+				}
+			}
+		),
+		onFocus: handle(
+			forward('onFocus'),
+			({target}) => showTooltip(target)
+		),
+		onBlur: handle(
+			forward('onBlur'),
+			hideTooltip
+		)
+	}, props);
+
+	useEffect(() => {
+		const mutableRefCurrent = mutableRef.current;
+
+		if (window.MutationObserver) {
+			mutableRefCurrent.mutationObserver = new MutationObserver(startTooltipLayoutJob);
+		}
+
+		if (window.ResizeObserver) {
+			mutableRefCurrent.resizeObserver = new ResizeObserver(startTooltipLayoutJob);
+		}
+
+		return () => {
+			if (currentTooltip === clientRef.current) {
+				currentTooltip = null;
+
+				mutableRefCurrent.mutationObserver?.disconnect();
+				mutableRefCurrent.resizeObserver?.disconnect();
+
+				mutableRefCurrent.showTooltipJob.stop();
+				mutableRefCurrent.setTooltipLayoutJob.stop();
+			}
+		};
+	}, [startTooltipLayoutJob]);
```

Comment:
Shouldn't creating and disconnecting `mutationObserver` and `resizeObserver` be for only mounting and unmounting?

Reply:
Well, not remember whole my thoughts during the work, I propabely add a dependency simply to turn off the warning (or pushed a button to update automatically by IDE).
Note that is is not a must-do thing to connect and disconnect only mounting and unmounting since the current code works safely.

Reply:
Concluded to keep the current dependency.